`Section#number` crashes for appendix sections

`Section#number` fails with:

```
SEVERE: org.jruby.RubyException

org.asciidoctor.internal.AsciidoctorCoreException: org.jruby.exceptions.RaiseException: (TypeError) asciidoctor: FAILED: <stdin>: Failed to load AsciiDoc document - cannot convert instance of class org.jruby.RubyString to int

	at org.asciidoctor.internal.JRubyAsciidoctor.render(JRubyAsciidoctor.java:307)
	at org.asciidoctor.internal.JRubyAsciidoctor.convert(JRubyAsciidoctor.java:512)
	at org.hibernate.infra.asciidoctor.extensions.customnumbering.ExampleNumberingTest.charNumberingFailsTest(ExampleNumberingTest.java:122)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:237)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
Caused by: org.jruby.exceptions.RaiseException: (TypeError) asciidoctor: FAILED: <stdin>: Failed to load AsciiDoc document - cannot convert instance of class org.jruby.RubyString to int
	at org.jruby.gen.InterfaceImpl1251850831.number(org/jruby/gen/InterfaceImpl1251850831.gen:13)
	at org.jruby.RubyMethod.call(org/jruby/RubyMethod.java:120)
	at RUBY.parse(/Users/marko/.m2/repository/org/asciidoctor/asciidoctorj/1.5.5/asciidoctorj-1.5.5.jar!/gems/asciidoctor-1.5.5/lib/asciidoctor/document.rb:479)
	at org.jruby.RubyArray.each(org/jruby/RubyArray.java:1613)
	at RUBY.parse(/Users/marko/.m2/repository/org/asciidoctor/asciidoctorj/1.5.5/asciidoctorj-1.5.5.jar!/gems/asciidoctor-1.5.5/lib/asciidoctor/document.rb:478)
	at RUBY.load(/Users/marko/.m2/repository/org/asciidoctor/asciidoctorj/1.5.5/asciidoctorj-1.5.5.jar!/gems/asciidoctor-1.5.5/lib/asciidoctor.rb:1347)
	at RUBY.convert(/Users/marko/.m2/repository/org/asciidoctor/asciidoctorj/1.5.5/asciidoctorj-1.5.5.jar!/gems/asciidoctor-1.5.5/lib/asciidoctor.rb:1465)
	at RUBY.convert(<script>:72)
	at org.jruby.gen.InterfaceImpl1900097566.convert(org/jruby/gen/InterfaceImpl1900097566.gen:13)
```

when called on appendix sections.

Here's a possible test to see the problem:

```java
public class TreeprocessorTest {

   private static Asciidoctor DOCTOR;

   @BeforeClass
   public static void initAsciidoctor() {
      DOCTOR = Asciidoctor.Factory.create();
      DOCTOR.javaExtensionRegistry().treeprocessor( CustomNumberingProcessor.class );
   }

   @Test
   public void charNumberingFailsTest() {
      DOCTOR.javaExtensionRegistry().treeprocessor( new Treeprocessor() {

         @Override public org.asciidoctor.ast.Document process(org.asciidoctor.ast.Document document) {
            Map<Integer, Section> sectionIndexMap = document.getBlocks().stream().filter( block -> block instanceof Section )
                  .map( block -> (Section) block )
                  .collect( Collectors.toMap( section -> section.numbered() ? section.number() : -1, section -> section ) );
            return document;
         }
      } );

      String content = "= Some title\n"
            + "\n"
            + ":sectnums:\n"
            + "== Some other title\n"
            + "\n"
            + "some very useful text which makes no sense.\n"
            + "\n"
            + "=== Some other inner title\n"
            + "\n"
            + "some very useful text which makes no sense.\n"
            + "\n"
            + "[appendix]\n"
            + "== Yet another title\n"
            + "\n"
            + "some very useful text which makes no sense.\n"
            + "\n"
            + "[appendix]\n"
            + "== Yet another title\n"
            + "\n"
            + "some very useful text which makes no sense.\n";
      String result = DOCTOR.convert( content, OptionsBuilder.options().safe( SafeMode.SAFE ).asMap() );
      Assert.assertNotNull( "result cannot be null", result );
      Assert.assertTrue( "converted result cannot be empty", !result.trim().isEmpty() );
   }
}
```

After looking into the problem - it seems that in Ruby `@number` can either be `fixnum` or `string`. So it seems that either adding class like `AppendixSection` (in which `AppendixSection#number` will return `String`) or somehow changing (or adding different method?) the `Section#number` is required to solve the issue.

--- a/asciidoctorj-api/src/main/java/org/asciidoctor/ast/Section.java
+++ b/asciidoctorj-api/src/main/java/org/asciidoctor/ast/Section.java
@@ -27,6 +27,16 @@ public interface Section extends StructuralNode {
     int getNumber();
 
     /**
+     * Section numeral.
+     *
+     * This is a roman numeral for book parts, arabic numeral for regular sections,
+     * and a letter for special sections.
+     *
+     * @return the section numeral
+     */
+    String getNumeral();
+
+    /**
      * @deprecated Please use {@link #getSectionName()}
      * @return the section name of this section
      */

--- a/asciidoctorj-api/src/main/java/org/asciidoctor/ast/Section.java
+++ b/asciidoctorj-api/src/main/java/org/asciidoctor/ast/Section.java
@@ -22,8 +22,10 @@ public interface Section extends StructuralNode {
     int number();
 
     /**
+     * @deprecated Please use {@link #getNumeral()}
      * @return the number of this section within the parent block
      */
+    @Deprecated
     int getNumber();
 
     /**

--- a/asciidoctorj-core/src/main/java/org/asciidoctor/jruby/ast/impl/SectionImpl.java
+++ b/asciidoctorj-core/src/main/java/org/asciidoctor/jruby/ast/impl/SectionImpl.java
@@ -32,6 +32,11 @@ public class SectionImpl extends StructuralNodeImpl implements Section {
     }
 
     @Override
+    public String getNumeral() {
+        return getString("numeral");
+    }
+
+    @Override
     @Deprecated
     public String sectname() {
         return getSectionName();

--- a/asciidoctorj-core/src/main/java/org/asciidoctor/jruby/ast/impl/SectionImpl.java
+++ b/asciidoctorj-core/src/main/java/org/asciidoctor/jruby/ast/impl/SectionImpl.java
@@ -27,6 +27,7 @@ public class SectionImpl extends StructuralNodeImpl implements Section {
     }
 
     @Override
+    @Deprecated
     public int getNumber() {
         return getInt("number");
     }

--- a/asciidoctorj-core/src/test/groovy/org/asciidoctor/WhenADocumentContainsNumberedSections.groovy
+++ b/asciidoctorj-core/src/test/groovy/org/asciidoctor/WhenADocumentContainsNumberedSections.groovy
@@ -1,0 +1,73 @@
+package org.asciidoctor
+
+import org.asciidoctor.ast.Section
+import org.jboss.arquillian.spock.ArquillianSputnik
+import org.jboss.arquillian.test.api.ArquillianResource
+import org.junit.runner.RunWith
+import spock.lang.Specification
+
+@RunWith(ArquillianSputnik)
+class WhenADocumentContainsNumberedSections extends Specification {
+
+    @ArquillianResource
+    private Asciidoctor asciidoctor
+
+    def "for appendix the section numeral should be a letter"() {
+        given:
+        final source = '''\
+            = Article Title
+            :sectnums:
+            
+            [appendix]
+            == Appendix Title
+            '''.stripIndent()
+
+        when:
+        def document = asciidoctor.load source, [:]
+
+        then:
+        final appendix = (Section) document.blocks[0]
+        appendix.title == 'Appendix Title'
+        appendix.numeral == 'A'
+    }
+
+    def "for regular section the numeral should be an arabic numeral"() {
+        given:
+        final source = '''\
+            = Article Title
+            :sectnums:
+            
+            == Regular Section Title
+            '''.stripIndent()
+
+        when:
+        def document = asciidoctor.load source, [:]
+
+        then:
+        final section = (Section) document.blocks[0]
+        section.title == 'Regular Section Title'
+        section.numeral == '1'
+    }
+
+    def "for book part the section numeral should be a roman numeral"() {
+        given:
+        final source = '''\
+            = Book Title
+            :doctype: book
+            :sectnums:
+            :partnums:
+            
+            = Part Title
+            
+            == Chapter Title
+            '''.stripIndent()
+
+        when:
+        def document = asciidoctor.load source, [:]
+
+        then:
+        final part = (Section) document.blocks[0]
+        part.title == 'Part Title'
+        part.numeral == 'I'
+    }
+}

